File: app/src/searchdolls.rb
require 'cgi'
require 'open-uri'

class DflSearch
  @@base_uri = 'https://cdn.wikiwiki.jp/to/w/dolls-fl/'

  def mk_url(doll_name, is_damage)
    file_name = doll_name
    file_name += '_damage' if is_damage
    url = "#{doll_name}/::ref/#{file_name}.jpg"
    return @@base_uri + url
  end

  def fetchable(url)
    begin
      open(url)
    rescue OpenURI::HTTPError
      return false
    end

    return true
  end

  def get_doll_pict(doll_name, is_damage)
    doll_name = CGI.escape(doll_name)
    doll_name.gsub!('+', '%20')
    url = mk_url(doll_name, is_damage)

    return '該当するドールが見つかりません' unless fetchable(url)
    $reply_type = 'image'
    return url
  end

  def answer(*msg_data)
    msg = msg_data[0]
    return unless msg =~ /^doll /

    is_damage = false
    doll_name = msg.sub(/^doll /, '')

    if doll_name =~ /damage/
      is_damage = true
      doll_name.sub!(/damage /, '')
    end
    get_doll_pict(doll_name, is_damage)
  end
end

File: gokabot-api-go/internal/answerers/searchdolls/searchdolls.go
package searchdolls

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/23prime/gokabot-api/internal/answerer"
)

const baseURL = "https://cdn.wikiwiki.jp/to/w/dolls-fl/"

// Answerer looks up Girls' Frontline doll images on cdn.wikiwiki.jp.
// Trigger: "doll <name>" or "doll damage <name>"
type Answerer struct {
	client  *http.Client
	baseURL string
}

func New() *Answerer {
	return &Answerer{
		client:  &http.Client{Timeout: 5 * time.Second},
		baseURL: baseURL,
	}
}

func (a *Answerer) Answer(data answerer.MessageData) *answerer.Response {
	msg := data.Message
	if !strings.HasPrefix(msg, "doll ") {
		return nil
	}

	dollName := strings.TrimPrefix(msg, "doll ")
	imageURL := a.buildURL(dollName)

	if !a.fetchable(imageURL) {
		return &answerer.Response{Text: "該当するドールが見つかりません", ReplyType: "text"}
	}
	return &answerer.Response{Text: imageURL, ReplyType: "image"}
}

// buildURL constructs the wikiwiki image URL for the given doll name.
// "damage <name>" appends "_damage" to the filename.
func (a *Answerer) buildURL(dollName string) string {
	isDamage := strings.HasPrefix(dollName, "damage ")
	if isDamage {
		dollName = strings.TrimPrefix(dollName, "damage ")
	}

	encoded := url.PathEscape(dollName)
	fileName := encoded
	if isDamage {
		fileName += "_damage"
	}

	return fmt.Sprintf("%s%s/::ref/%s.jpg", a.baseURL, encoded, fileName)
}

func (a *Answerer) fetchable(imageURL string) bool {
	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, imageURL, nil) //nolint:gosec
	if err != nil {
		return false
	}

	resp, err := a.client.Do(req)
	if err != nil {
		return false
	}
	defer resp.Body.Close() //nolint:errcheck

	return resp.StatusCode < 400
}
